ext/surro-gate/selector_ext.c

```diff
@@ -164,29 +135,23 @@ static VALUE SurroGate_Selector_select(VALUE self, VALUE timeout) {
   rb_thread_call_without_gvl(wait_func, &wait_args, NULL, NULL);
 
   for (i=0; i<wait_args.result; i++) {
-    source = (int)((events[i].data.u64 & 0xFFFFFFFF00000000LL) >> 32);
-    target = (int)(events[i].data.u64 & 0xFFFFFFFFLL);
-
-    read = rb_funcall(pairing, rb_intern("[]"), 1, INT2NUM(target)); // @pairing[target]
-    write = rb_funcall(pairing, rb_intern("[]"), 1, INT2NUM(source)); // @pairing[source]
+    socket = (VALUE) events[i].data.u64;
 
     if (events[i].events & EPOLLIN && events[i].events & EPOLLOUT) {
       // Socket is both available for read and write
-      rb_iv_set(read, "@rd_rdy", Qtrue); // read.rd_rdy = true
-      rb_iv_set(write, "@wr_rdy", Qtrue); // write.wr_rdy = true
+      rb_funcall(scoreboard, rb_intern("mark_rd"), 1, socket);
+      rb_funcall(scoreboard, rb_intern("mark_wr"), 1, socket);
     } else if (events[i].events & EPOLLIN) {
       // Socket is available for read, reregister it for write if not writable
-      rb_iv_set(read, "@rd_rdy", Qtrue); // read.rd_rdy = true
-      if (rb_iv_get(write, "@wr_rdy") == Qfalse) { // if !write.wr_rdy
-        socket = rb_iv_get(read, "@left"); // read.left
-        epoll_rearm(selector, SOCK_PTR(socket), target, source, EPOLLOUT);
+      rb_funcall(scoreboard, rb_intern("mark_rd"), 1, socket);
+      if (rb_funcall(scoreboard, rb_intern("marked_wr?"), 1, socket) == Qfalse) {
+        epoll_rearm(selector, socket, EPOLLOUT);
       }
     } else if (events[i].events & EPOLLOUT) {
       // Socket is available for write, reregister it for read if not readable
-      rb_iv_set(write, "@wr_rdy", Qtrue); // write.wr_rdy = true
-      if (rb_iv_get(read, "@rd_rdy") == Qfalse) { // if !source.rd_rdy
-        socket = rb_iv_get(write, "@right"); // write.right
-        epoll_rearm(selector, SOCK_PTR(socket), source, target, EPOLLIN);
+      rb_funcall(scoreboard, rb_intern("mark_wr"), 1, socket);
+      if (rb_funcall(scoreboard, rb_intern("marked_rd?"), 1, socket) == Qfalse) {
```

Comment:
feels like this could be a single ruby method call here (and the mark_rd ; marked_wr?)

Having trouble with name something like `mark_re_can_wr?`

to be honest, I'd imaging this is the most comon case for accessing this variable

Reply:
I'd rather keep this logic in the C code...